src/core/server/plugins/discovery/plugin_manifest_parser.ts

```diff
@@ -0,0 +1,174 @@
+ * directory path and produces an error result if it fails to do so or plugin manifest
+ * isn't valid.
+ * @param pluginPath Path to the plugin directory where manifest should be loaded from.
+ * @param packageInfo Kibana package info.
+ */
+export function parseManifest$(pluginPath: string, packageInfo: PackageInfo) {
+  const manifestPath = resolve(pluginPath, MANIFEST_FILE_NAME);
+  return fsReadFile$(manifestPath).pipe(
+    catchError(err => throwError(PluginDiscoveryError.missingManifest(manifestPath, err))),
+    map(manifestContent => {
+      let manifest: Partial<PluginManifest>;
+      try {
+        manifest = JSON.parse(manifestContent.toString());
+      } catch (err) {
+        throw PluginDiscoveryError.invalidManifest(manifestPath, err);
+      }
+
+      if (!manifest || typeof manifest !== 'object') {
+        throw PluginDiscoveryError.invalidManifest(
+          manifestPath,
+          new Error('Plugin manifest must contain a JSON encoded object.')
+        );
+      }
+
+      if (!manifest.id || typeof manifest.id !== 'string') {
+        throw PluginDiscoveryError.invalidManifest(
+          manifestPath,
+          new Error('Plugin manifest must contain an "id" property.')
+        );
+      }
+
+      if (!manifest.version || typeof manifest.version !== 'string') {
+        throw PluginDiscoveryError.invalidManifest(
+          manifestPath,
+          new Error(`Plugin manifest for "${manifest.id}" must contain a "version" property.`)
+        );
+      }
+
+      const expectedKibanaVersion =
+        typeof manifest.kibanaVersion === 'string' && manifest.kibanaVersion
+          ? manifest.kibanaVersion
+          : manifest.version;
+      if (!isVersionCompatible(expectedKibanaVersion, packageInfo.version)) {
+        throw PluginDiscoveryError.incompatibleVersion(
+          manifestPath,
+          new Error(
+            `Plugin "${
+              manifest.id
+            }" is only compatible with Kibana version "${expectedKibanaVersion}", but used Kibana version is "${
+              packageInfo.version
+            }".`
+          )
+        );
+      }
+
+      return {
+        id: manifest.id,
+        version: manifest.version,
+        kibanaVersion: expectedKibanaVersion,
+        requiredPlugins: Array.isArray(manifest.requiredPlugins) ? manifest.requiredPlugins : [],
+        optionalPlugins: Array.isArray(manifest.optionalPlugins) ? manifest.optionalPlugins : [],
+        ui: typeof manifest.ui === 'boolean' ? manifest.ui : false,
+      };
+    })
+  );
+}
+
+/**
+ * Checks whether plugin expected Kibana version is compatible with the used Kibana version.
+ * @param expectedKibanaVersion Kibana version expected by the plugin.
+ * @param actualKibanaVersion Used Kibana version.
+ */
+function isVersionCompatible(expectedKibanaVersion: string, actualKibanaVersion: string) {
+  if (expectedKibanaVersion === ALWAYS_COMPATIBLE_VERSION) {
```

Comment:
__note:__ I brought this from the legacy world, but not 100% sure we still want to keep that in the new platform, I'd say `yes`, but what do you think?

Reply:
👍 I think we should keep it for now